### PR TITLE
feat: taints & toleration plugin

### PR DESCRIPTION
**Pull request by Arvindthiru**
### Description of your changes

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### How has this code been tested

UTs

ITs and E2Es will be added in separate PRs

### CODE REVIEW

File: pkg/scheduler/framework/framework.go [hunk 1]
```diff
@@ -128,9 +135,13 @@ type frameworkOptions struct {
...
 	maxUnselectedClusterDecisionCount int
 
-	// checker is the cluster eligibility checker the scheduler framework will use to check
+	// clusterEligibilityChecker is the checker the scheduler framework will use to check
 	// if a cluster is eligibile for resource placement.
 	clusterEligibilityChecker *clustereligibilitychecker.ClusterEligibilityChecker
+
+	// taintTolerationChecker is the checker the scheduler framework will use to check
+	// if taints on a member cluster can be tolerated by tolerations on cluster resource placement.
+	taintTolerationChecker *tainttolerationchecker.TaintTolerationChecker
```
Comment (Arvindthiru) [inline pkg/scheduler/framework/framework.go:144]: This checker will be used for PickFixed placement type similar to clusterEligibilityChecker where we don't run all the filter plugins

File: pkg/scheduler/framework/plugins/tainttoleration/filtering.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+// Filter allows the plugin to connect to the Filter extension point in the scheduling framework.
+func (p *Plugin) Filter(
+	_ context.Context,
+	_ framework.CycleStatePluginReadWriter,
+	policy *placementv1beta1.ClusterSchedulingPolicySnapshot,
+	cluster *clusterv1beta1.MemberCluster,
+) (status *framework.Status) {
+	taint, isUntolerated := p.handle.TaintTolerationChecker().FindUnToleratedTaint(cluster.Spec.Taints, utils.GetTolerations(policy.Spec.Policy))
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/filtering.go:24]: what if the cluster does not have any taint?

File: pkg/utils/common.go [hunk 3]
```diff
@@ -391,3 +391,11 @@ func GenerateGroupString(groups []string) string {
 	}
 	return groupString
 }
+
+// GetTolerations returns tolerations from supplied placement policy.
+func GetTolerations(policy *placementv1beta1.PlacementPolicy) []placementv1beta1.Toleration {
```
Comment (zhiying-lin) [inline pkg/utils/common.go:396]: prefer to move this util into apis/placement package, something like,

func (m *ClusterResourcePlacement) Tolerations() []Toleration  {
	....
}

similar to policySnapshot.

We need to get rid of the "utils" package cause the package name is not good and does not provide any meaning.

File: pkg/utils/test_util.go [hunk 4]
```diff
@@ -136,3 +142,32 @@ func (matcher AlreadyExistMatcher) FailureMessage(actual interface{}) (message s
 func (matcher AlreadyExistMatcher) NegatedFailureMessage(actual interface{}) (message string) {
 	return format.Message(actual, "not to be already exist")
 }
+
+// MockHandle framework.Handle interface for set up the plugin.
+type MockHandle struct {
```
Comment (zhiying-lin) [inline pkg/utils/test_util.go:147]: All the test util functions need to be moved to the test/utils/ path. We should not mix the test util with the non test utils in one package.

File: pkg/scheduler/framework/framework.go (hunk 1 above)
Comment (Arvindthiru) [reply pkg/scheduler/framework/framework.go]: Discussed offline taints and tolerations plugin will not be run for pickFixed since it's similar to specifying nodeName in podSpec

File: pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go [hunk 5]
```diff
@@ -0,0 +1,410 @@
...
+				Spec: placementv1beta1.SchedulingPolicySnapshotSpec{
+					Policy: &placementv1beta1.PlacementPolicy{
+						PlacementType: placementv1beta1.PickAllPlacementType,
+					},
+				},
+			},
+			wantStatus: framework.NewNonErrorStatus(framework.ClusterUnschedulable, p.Name(), fmt.Sprintf(reasonFmt, &clusterv1beta1.Taint{Key: "key1", Value: "value1", Effect: corev1.TaintEffectNoSchedule})),
+		},
+		{
+			name: "taints can be tolerated based on key, value & effect with Equal operator, every taint has matching toleration - nil status",
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go:82]: add some (positive and negative) tests when taint key is present + empty value + operation is equal.

File: pkg/scheduler/framework/plugins/tainttoleration/plugin.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+// WithName sets the name of the plugin.
+func WithName(name string) Option {
+	return func(o *taintTolerationPluginOptions) {
+		o.name = name
+	}
+}
+
+// New returns a new Plugin.
+func New(opts ...Option) Plugin {
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/plugin.go:45]: you will add this plugin in your next PR?

Comment (Arvindthiru) [reply pkg/scheduler/framework/plugins/tainttoleration/plugin.go:45]: Yes

File: pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go [hunk 7]
```diff
@@ -171,7 +171,48 @@ func TestFilter(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "csp-1",
+				},
+				Spec: placementv1beta1.SchedulingPolicySnapshotSpec{
+					Policy: &placementv1beta1.PlacementPolicy{
+						PlacementType: placementv1beta1.PickAllPlacementType,
+						Tolerations: []placementv1beta1.Toleration{
+							{
+								Key:      "key1",
+								Operator: corev1.TolerationOpEqual,
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go:202]: it's not a valid test case? i remember this crp will be rejected by webhook when the value is empty?

File: pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go [hunk 8]
```diff
@@ -171,7 +171,48 @@ func TestFilter(t *testing.T) {
...
+							},
+							{
+								Key:      "key2",
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go:205]: add a test case when the operation is exist?

File: pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go [hunk 9]
```diff
@@ -0,0 +1,446 @@
...
+								Value:    "value2",
+							},
+						},
+					},
+				},
+			},
+			wantStatus: nil,
+		},
+		{
+			name: "taints can be tolerated based on key, empty value, effect with Equal operator, every taint has matching toleration - nil status",
```
Comment (zhiying-lin) [inline pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go:174]: add a negative test case for taints (key present + empty value + operation is equal)?

File: pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go (hunk 7 above)
Comment (Arvindthiru) [reply pkg/scheduler/framework/plugins/tainttoleration/filtering_test.go:202]: closed a fix PR to address this https://github.com/Azure/fleet/pull/704 to align with k8s
